tools/inventory-downloader/src/main.go

```diff
@@ -0,0 +1,211 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/feature/s3/manager"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"gopkg.in/confluentinc/confluent-kafka-go.v1/kafka"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"os/signal"
+	"path"
+	"strconv"
+	"syscall"
+)
+
+func extractFilenameFromURL(urlString string) (string, error) {
+	parsedURL, err := url.Parse(urlString)
+	if err != nil {
+		return "", err
+	}
+
+	filename := path.Base(parsedURL.Path)
+
+	return filename, nil
+}
+
+func downloadFile(url string) (io.ReadCloser, error) {
+	response, err := http.Get(url)
+	if err != nil {
+		return nil, err
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("HTTP request failed with status: %s", response.Status)
+	}
+
+	return response.Body, nil
+}
+
+func uploadToS3(cfg aws.Config, ctx context.Context, reader io.Reader, bucketName, objectKey string) error {
+	client := s3.NewFromConfig(cfg)
+	uploader := manager.NewUploader(client)
+
+	_, err := uploader.Upload(ctx, &s3.PutObjectInput{
+		Bucket: &bucketName,
+		Key:    &objectKey,
+		Body:   reader,
+	})
+	return err
+}
+
+func processFile(cfg aws.Config, ctx context.Context, objectKey string, fileUrl string) error {
+	reader, err := downloadFile(fileUrl)
+	if err != nil {
+		return fmt.Errorf("Error downloading file: %w", err)
+
+	}
+	defer reader.Close()
+
+	err = uploadToS3(cfg, ctx, reader, *bucketName, objectKey)
```

Comment:
I wonder if we can use S3 CopyObject functionality here? I am not sure if that works with pre-signed URLs, but it if does, it will be more efficient than downloading and uploading.